starknet/l2_minter/src/ethereum_minter.cairo

```diff
@@ -0,0 +1,252 @@
+#[starknet::contract]
+mod EthereumMinter {
+    // Library Imports
+    use starknet::class_hash::ClassHash;
+    use zeroable::Zeroable;
+    use starknet::ContractAddress;
+    use starknet::get_block_timestamp;
+    use core::option::OptionTrait;
+    use result::ResultTrait;
+    use traits::{Into, TryInto};
+
+    // use ethereum_minter::helpers::booking_storage::StorageAccessBooking;
+    use ethereum_minter::helpers::status::mint_status_to_u8;
+    use ethereum_minter::helpers::status::u8_to_mint_status;
+    // use ethereum_minter::interfaces::l1_minter::IEthereumMinter;
+    use ethereum_minter::interfaces::erc3525::{IERC3525Dispatcher, IERC3525DispatcherTrait};
+
+    #[derive(storage_access::StorageAccess, Drop, Copy)]
+    struct Booking {
+        value: u256,
+        amount: u256,
+        status: u8,
+    }
+
+    #[derive(storage_access::StorageAccess, Drop)]
+    enum MintStatus {
+        Booked: (),
+        Failed: (),
+        Minted: (),
+        Refunded: (),
+    }
+
+    #[storage]
+    struct Storage {
+        _l1_minter_address: felt252,
+        _l1_mint_counts: LegacyMap::<ContractAddress, u32>,
+        // booked_values: (user_address, user_mint_index) -> (value, amount, status)
+        _booked_values: LegacyMap::<(ContractAddress, u32), Booking>,
+        _projects_contract: IERC3525Dispatcher,
+        _slot: u256,
+        _unit_price: u256,
+        _max_supply: u256,
+        _max_value_per_tx: u256,
+        _min_value_per_tx: u256,
+        _current_max_supply: u256,
+    }
+
+    // Events
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        Upgraded: Upgraded,
+        BookingClaimed: BookingClaimed,
+        BookingHandled: BookingHandled,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct Upgraded {
+        implementation: ClassHash
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct BookingHandled {
+        address: ContractAddress,
+        value: u256,
+        time: u64
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct BookingClaimed {
+        address: ContractAddress,
+        value: u256,
+    }
+
+    // Methods
+    #[constructor]
+    fn init(
+        ref self: ContractState,
+        projects_contract: ContractAddress,
+        slot: u256,
+        unit_price: u256,
+        max_supply: u256,
+        max_value_per_tx: u256,
+        min_value_per_tx: u256,
+    ) {
+        assert(!projects_contract.is_zero(), 'Projects contract cannot be 0');
+        assert(unit_price > 0, 'Unit price must be positive');
+        assert(max_supply > 0, 'Max supply must be positive');
```

Comment:
Should `Max supply` be compared to ERC3525.getProjectValue ?

Reply:
check added